src/Components/src/Microsoft.AspNetCore.Components.Server/Rendering/HtmlRenderer.cs

```diff
@@ -0,0 +1,207 @@
+        /// of the HTML produced by the component.
+        /// </summary>
+        /// <typeparam name="T">The type of the <see cref="IComponent"/>.</typeparam>
+        /// <param name="initialParameters">A <see cref="ParameterCollection"/> with the initial parameters to render the component.</param>
+        /// <returns>A sequence of <see cref="string"/> fragments that represent the HTML text of the component.</returns>
+        public IEnumerable<string> RenderComponent<T>(ParameterCollection initialParameters) where T : IComponent
+        {
+            return RenderComponent(typeof(T), initialParameters);
+        }
+
+        /// <summary>
+        /// Renders a component into a sequence of <see cref="string"/> fragments that represent the textual representation
+        /// of the HTML produced by the component.
+        /// </summary>
+        /// <param name="componentType">The type of the <see cref="IComponent"/>.</param>
+        /// <param name="initialParameters">A <see cref="ParameterCollection"/> with the initial parameters to render the component.</param>
+        /// <returns>A sequence of <see cref="string"/> fragments that represent the HTML text of the component.</returns>
+        private IEnumerable<string> RenderComponent(Type componentType, ParameterCollection initialParameters)
+        {
+            var tree = CreateInitialRender(componentType, initialParameters);
+
+            var frames = tree.GetFrames();
+
+            if (frames.Count == 0)
+            {
+                return Array.Empty<string>();
+            }
+            else
+            {
+                var (segments, newPosition) = RenderFrames(frames, 0, frames.Count);
+                Debug.Assert(newPosition == frames.Count);
+                return segments;
+            }
+        }
+
+        private (IEnumerable<string>, int newPosition) RenderFrames(ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            IEnumerable<string> segments = Array.Empty<string>();
+            IEnumerable<string> nextSegments;
+            var nextPosition = position;
+            var endPosition = position + maxElements;
+            while (position < endPosition)
+            {
+                (nextSegments, nextPosition) = RenderCore(frames, position, maxElements);
+                if (position == nextPosition)
+                {
+                    throw new InvalidOperationException("We didn't consume any input.");
+                }
+                position = nextPosition;
+                segments = segments.Concat(nextSegments);
+            }
+
+            return (segments, nextPosition);
+        }
+
+        private (IEnumerable<string> segments, int nextPosition) RenderCore(
+            ArrayRange<RenderTreeFrame> frames,
+            int position,
+            int length)
+        {
+            var frame = frames.Array[position];
```

Comment:
I think this is going to copy the `RenderTreeFrame` struct. To avoid that, is it safe to use `ref var frame = ref frames.Array[position];` here?

Reply:
Hmm, probably. I focused on writing the algorithm for clarity first (It was also part of me figuring things out) and now I'm going to do an optimization pass. I've never used ref locals but I'll give it a try.